### PR TITLE
Feat: Add product variant rank

### PR DESCRIPTION
**Pull request by pKorsholm**
**What?**
- Add column "variant_rank" to product-variant in the database
- Sort product variants according to rank when retrieving a product
- Update variant rankings when updating a product
- Assign variant ranking when creating a variant (appended to end of variants)
- Add service, api and integration tests
- update product-variant db table with cascading deletes (both sides of relation to money-amount)

**Why?**
- allows admins to reorder product variants 
- Deleting a variant with money-amounts correctly deletes the variant and money-amount 

**How?**
- Update rank: iterate variants on update assigning the index of each variant to its rank
- fetch product: sort variants based on rank before returning from the api

**Testing**
- Add unit testing of services and api to test rankings 
- Add integration tests
- Integration tests
  - GET /admin/products
	- _returns a list of products with child entities_: tests that the list-products endpoint correctly returns a product including the nested entities, this was added after moving sorting of variants to the database which required switching to a querybuilder,

  - POST /admin/products 
	- _Sets variant ranks when creating a product_: tests that the variants are returned in the order they were inserted in
	- _Updates a product (variant ordering)_: tests that updating a product with variants in a different order updates the ranks of the variants

### CODE REVIEW

File: packages/medusa/src/services/product.js [hunk 1]
```diff
@@ -185,6 +185,12 @@ class ProductService extends BaseService {
       )
     }
 
+    if (product.variants) {
+      product.variants.sort(
+        (variant1, variant2) => variant1.variant_rank - variant2.variant_rank
+      )
+    }
+
```
Comment (srindom) [inline packages/medusa/src/services/product.js:193]: Can we do the sorting at the database level instead? Potentially in `findOneWithRelations`

File: packages/medusa/src/services/product.js [hunk 2]
```diff
@@ -293,6 +299,10 @@ class ProductService extends BaseService {
         rest.discountable = false
       }
 
+      if (rest.variants)
+        for (const [i, variant] of rest.variants.entries())
+          variant.variant_rank = i
+
```
Comment (srindom) [inline packages/medusa/src/services/product.js:305]: variants are not created as part of the `create` call here - this happens in the route which you are already accounting for so please delete this.

File: packages/medusa/src/migrations/1630868939283-product_variant_rank.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import {MigrationInterface, QueryRunner} from "typeorm";
+
+export class productVariantRank1630868939283 implements MigrationInterface {
+    name = 'productVariantRank1630868939283'
+
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(`ALTER TABLE "product_variant" ADD "variant_rank" integer NOT NULL`);
```
Comment (srindom) [inline packages/medusa/src/migrations/1630868939283-product_variant_rank.ts:7]: the `NOT NULL` constraint here will break the migration for existing `product_variants` as these will have null values for the column by default - would simply leave out the NOT NULL constraint and add a default value of 0 instead

**Review by srindom — CHANGES_REQUESTED** [review]
Nice work - a couple of fixes to get out of the way here.

Would also be great if you could break down each of the test cases you account for in your integration tests when writing the PR description :)

File: packages/medusa/src/services/product.js (hunk 2 above)
Comment (pKorsholm) [reply packages/medusa/src/services/product.js]: I removed it

File: packages/medusa/src/migrations/1631104895519-RankColumnWithDefaultValue.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import {MigrationInterface, QueryRunner} from "typeorm";
```
Comment (srindom) [inline packages/medusa/src/migrations/1631104895519-RankColumnWithDefaultValue.ts:1]: Please update the filename to be snake_cased like all the other migrations -> https://github.com/medusajs/medusa/tree/master/packages/medusa/src/migrations

File: packages/medusa/src/services/__mocks__/product.js [hunk 5]
```diff
@@ -36,6 +36,8 @@ export const ProductServiceMock = {
     if (data.title === "Test Product") {
       return Promise.resolve(products.product1)
     }
+    if (data.title === "Test Product with variants")
+      return Promise.resolve(products.productWithOptions)
```
Comment (srindom) [inline packages/medusa/src/services/__mocks__/product.js:41]: Use braces for all control structures: https://github.com/airbnb/javascript#blocks

**Review by srindom — CHANGES_REQUESTED** [review]
nice work!! small fixes and we should be good to go 🎉

**Review by srindom — APPROVED** [review]
Let's go! 🚀

**Pull request merged**